service/lib/agama/software/callbacks/provide.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+# Copyright (c) [2025] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "agama/question"
+
+Yast.import "Pkg"
+
+module Agama
+  module Software
+    module Callbacks
+      # Provide callbacks
+      class Provide
+        include Yast::I18n
+
+        # From https://github.com/openSUSE/libzypp/blob/d90a93fc2a248e6592bd98114f82a0b88abadb72/zypp/ZYppCallbacks.h#L111
+        NO_ERROR = 0
+        NOT_FOUND = 1
+        IO_ERROR = 2
+        INVALID = 3
+
+        # Constructor
+        #
+        # @param questions_client [Agama::DBus::Clients::Questions]
+        # @param logger [Logger]
+        def initialize(questions_client, logger)
+          @questions_client = questions_client
+          @logger = logger
+        end
+
+        # Register the callbacks
+        def setup
+          Yast::Pkg.CallbackDoneProvide(
+            Yast::FunRef.new(method(:done_provide), "string (integer, string, string)")
+          )
+        end
+
+        # DoneProvide callback
+        #
+        # @return [String] "I" for ignore, "R" for retry and "C" for abort (not implemented)
```

Comment:
NP: I think it would be better to use constants like `IGNORE` and `RETRY` in code instead of C or I.